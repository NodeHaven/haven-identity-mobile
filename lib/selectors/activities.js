// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { createSelector } from 'reselect'
import { get, toJs, map, vals, reverse, take, filter, remove, count, vector, into } from 'mori'
import { currentIdentity, selectedIdentity } from './identities'

// Filters a transaction by the 'to' field or 'client_id' field
export const byAddress = (state, address) => {
  return filter((act) => (get(act, 'to') === address || get(act, 'client_id') === address), activities(state))
}

// Returns a boolean if the activity is unread
// Used as a filter in other selectors
const isUnread = (activity) => {
  return !(get(activity, 'opened') || get(activity, 'authorizedAt') || get(activity, 'canceledAt'))
}

const unopenedAttestations = (activity) => (
  get(activity, 'type') === 'attestation' && !get(activity, 'opened')
)
const activeNotification = (activity) => {
  return (!(get(activity, 'authorizedAt') || get(activity, 'canceledAt'))) || unopenedAttestations(activity)
}

// Returns Mori data structure of all activities for the current identity
// Returns an array if no activities are present
export const activities = createSelector(
  [currentIdentity],
  (identity) => reverse(filter(a => get(a, 'id'), vals(get(identity, 'activities')))) || []
)

export const notifications = createSelector(
  [activities],
  (rawActivities) => into(vector(), filter(activeNotification, rawActivities))
)

// Returns an integer detailing the current count of unread activities
export const notificationCount = createSelector(
  [notifications],
  (_activities) => count(_activities)
)

// Returns aMori data structure of all activities for a selectedIdentity
// Returns an array if no activities are present
export const activitiesForAddress = createSelector(
  [selectedIdentity],
  (identity) => reverse(filter(a => get(a, 'id'), vals(get(identity, 'activities')))) || []
)

// Returns a JS Object of activities sorted by address
// Return an empty array if nothing is found
export const activitiesByAddress = createSelector(
  [byAddress], (activities) => toJs(activities)
)

// Returns a JS Object of unconfirmedTransactions for an address
export const unconfirmedTransactions = createSelector(
  [activitiesForAddress], (all) => toJs(
    map((tx) => [get(tx, 'id'), get(tx, 'txhash')],
      filter(a => get(a, 'txhash') && !get(a, 'blockNumber'), all)))
)

// Returns a Mori data structure of all unread activities
// Returns an empty array if no activities are present
export const rawUnread = createSelector(
  [activities],
  (_activities) => filter(isUnread, _activities)
)

// Returns a JS Object of the first 10 unread activities
export const unread = createSelector(
  [rawUnread],
  (_activities) => toJs(take(10, _activities))
)

// Returns an integer detailing the current count of unread activities
export const unreadCount = createSelector(
  [rawUnread],
  (_activities) => count(_activities)
)

// Returns a Mori data structure of all activities that have been read
export const rawHistory = createSelector(
  [activities],
  (_activities) => remove(isUnread, _activities)
)

// Returns a JS Object of the first 30 read activities
export const history = createSelector(
  [rawHistory],
  (_activities) => toJs(take(30, _activities))
)

// Returns an integer detailing the current count of all activities minus unread
// Returns 0 if no activities exist
export const historyCount = createSelector(
  [rawHistory],
  (_activities) => _activities ? count(_activities) : 0
)
