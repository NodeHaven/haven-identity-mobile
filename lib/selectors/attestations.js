// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { createSelector } from 'reselect'
import { isEmpty, toJs, merge, toClj, hashMap, hasKey, count, selectKeys, vector, assoc, flatten, get, getIn, vals, map, reduce, reduceKV, inc, updateIn, sortedMap, first, groupBy, keys, dissoc, filter, sortBy, concat, conj } from 'mori'
import { currentIdentity, ownClaims, currentDidObject, externalIdentities as external } from './identities'
import { allProfiles } from './vc'

// Cut of for fixing JWT timestamps incorrectly set using milliseconds
export const JWT_MS_CUTOFF = 1451628000000

function normalizeTime (secsOrMs) {
  return secsOrMs > JWT_MS_CUTOFF ? Math.floor(secsOrMs / 1000) : secsOrMs
}

const externalIdentities = createSelector(
  [external, allProfiles],
  (ext, vcs) => vcs.reduce((all, profile) => updateIn(all, [profile.address], p => merge(p, toClj(profile))), ext || hashMap())
)

// NEEDS ERROR HANDLING
// Props needs to have the following values:
// - `issuer` an address
export const issuedBy = (state, props) => {
  return filter((att) => get(att, 'iss') === props.issuer, attestationLedger(state))
}

const claimsForSpec = (spec, claims) => {
  const claimType = typeof spec === 'object' ? spec.type : spec
  if (typeof claimType !== 'string') return []

  const candidates = map(claim => assoc(claim, 'claimType', claimType), filter(claim => hasKey(get(claim, 'claim'), claimType), claims))
  if (typeof spec === 'object' && spec.iss) {
    const iss = spec.iss.map ? spec.iss.map(spec => spec.did) : [spec.iss]
    return filter(claim => iss.includes(get(claim, 'iss')), candidates)
  } else {
    return candidates
  }
}
export const requestedAttestations = (state, requested) => {
  const all = onlyLatestAttestations(state)
  return flatten(map(spec => claimsForSpec(spec, all), requested))
}

const requested = (state, params) => params

const forIssuerAndClaimType = (state, iss, claimType) => ({iss, claimType})

// Props needs to have the following values:
// - `claimType` (eg. `name` or `phone`)
// - `claimValue` (eg. `Alice` ....)

export const forTypeAndValue = (state, props) => {
  return filter((att) => getIn(att, ['claim', props.claimType]) === props.claimValue, onlyLatestAttestations(state))
}

// This represents an array of claim objects
// { sub: '0x88647b5a94cd8347c8f76c7b79c85bbfbf0e13a9',
// claim: { employer: 'Consensys AG' },
// iss: '0x88647b5a94cd8347c8f76c7b79c85bbfbf0e13a9',
// iat: 1482268217248,
// exp: 1482354617248,
// token: 'tokenstring' }

// Return a mori object of the attestations for the current identity
export const attestationLedger = createSelector(
  [currentIdentity, currentDidObject],
 (identity, did) => sortBy(a => normalizeTime(get(a, 'iat')), (a, b) => b - a, concat(vals(get(identity, 'attestations')), vals(get(did, 'attestations'))))
// (identity, did) => concat(reverse(vals(get(identity, 'attestations'))), reverse(vals(get(did, 'attestations'))))
)

// Return a mori object of the most recent attestations for the current identity
export const onlyLatestAttestations = createSelector(
  [attestationLedger],
  (attestations) => map(first, vals(groupBy(a => vector(get(a, 'iss'), get(a, 'claim')), attestations)))
)

// Return a mori object with the following structure
// This represents an object that has keys for the attestations and an object
// with values and the number of attestations for each attestation
// { employer: { 'Consensys AG': 5 } }
export const thirdPartyAttestationSummary = createSelector(
  [onlyLatestAttestations],
  (attestations) =>
    reduce(
      (summary, att) => {
        // TODO this only deals with a single entry claims map. It should handle multiples
        const claim = get(att, 'claim')
        const claimType = first(keys(claim))
        return updateIn(summary, [claimType, get(claim, claimType)], inc)
      },
      sortedMap(),
      attestations
    )
)

// Return a mori object with the follow structure
// This represents a summary of all attestations with included claims
// { employer: { 'Consensys AG': 5 },
//   name: { 'Ashoka Finley': 0 },
//   phone: { '13104041586': 0 } }
export const attestationSummaryStats = createSelector(
  [thirdPartyAttestationSummary, ownClaims],
  (attestations, own) =>
    reduceKV((summary, type, value) => updateIn(summary, [type, value], (amount) => amount || 0),
          attestations,
          dissoc(own, 'connections'))
)

// Return a JS object from the attestationSummaryStats
export const attestationSummary = createSelector(
  [attestationSummaryStats],
  (summary) => toJs(summary))

// Return a mori object with all the attestations for the current identity
export const allClaims = createSelector(
  [attestationSummaryStats],
  summary => reduceKV((m, type, value) => assoc(m, type, first(keys(value))), hashMap(), summary)
)

// Return a JS object of the requested claims
// Return an empty object if none are present
export const requestedClaims = createSelector(
  [allClaims, requested],
  (claims, params) => toJs(selectKeys(claims, map(claimType => typeof claimType === 'object' ? claimType.type : claimType, params)))
)

// Return a JS object of the requested claims
// Return an empty object if none are present
export const requestedOwnClaims = createSelector(
  [ownClaims, requested],
  (claims, params) => toJs(selectKeys(claims, params))
)

// NEEDS ERROR HANDLING
// Returns a js object for typeAndValue for only the latest attestations
// Returns an empty array if none are present
export const attestationsForTypeAndValue = createSelector(
  [forTypeAndValue], (attestations) => toJs(attestations)
)

// Not Used
export const attestationsIssuedBy = createSelector(
  [issuedBy], (attestations) => toJs(attestations)
)

// Not Used
export const verifiedClaims = createSelector(
  [requestedAttestations],
  (attestations) => toJs(attestations)
)

// Returns a JS object of tokens of verifiedClaims
// Returns an empty object if none are found
export const verifiedClaimsTokens = createSelector(
  [requestedAttestations],
  (attestations) => toJs(map(a => get(a, 'token'), attestations))
)

// Not Used
export const verifiedClaimsByClaim = createSelector(
  [requestedAttestations],
  (attestations) => toJs(groupBy(a => get(a, 'claim'), attestations))
)

export const requestedWithIssuer = createSelector(
  [requestedAttestations, externalIdentities],
  (attestations, issuers) => {
    return map(att => assoc(att, 'issuer', get(issuers, get(att, 'iss'), {})), attestations)
  }
)

export const requestedVerifiableClaims = createSelector(
  [requestedWithIssuer],
  attestations => toJs(attestations)
)

export const missingClaims = createSelector(
  [requestedWithIssuer, requested],
  (attestations, requested) => requested.map(spec => typeof spec === 'object' ? spec : {type: spec}).filter(spec => isEmpty(filter(vc => get(vc, 'claimType') === spec.type, attestations)))
)

export const verifiedClaimsByType = createSelector(
  [requestedWithIssuer],
  (attestations) => toJs(groupBy(a => first(keys(get(a, 'claim'))), attestations))
)

export const attestationsForIssuerAndType = createSelector(
  [onlyLatestAttestations, forIssuerAndClaimType],
  (attestations, {iss, claimType}) => filter((att) => get(att, 'iss') === iss && getIn(att, ['claim', claimType]), attestations)
)

export const onlyLatestAttestationsWithIssuer = createSelector(
  [onlyLatestAttestations, externalIdentities],
  (latest, issuers) => toJs(map(
    att => assoc(att, 'issuer', get(issuers, get(att, 'iss'), {})),
    latest))
)

export const hasAttestations = createSelector(
  [onlyLatestAttestations],
  attestations => !isEmpty(attestations)
)
