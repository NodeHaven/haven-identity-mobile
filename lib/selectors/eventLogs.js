// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { createSelector } from 'reselect'
import { get, toJs } from 'mori'
import BN from 'bn.js'
const contractEvents = (state, address) => get(state.events, address)

export const eventBlock = createSelector(
  [contractEvents],
  (contract) => get(contract, 'block') ? new BN(get(contract, 'block'), 16) : undefined
)

// Not Used
export const eventLog = createSelector(
  [contractEvents],
  (contract) => toJs(get(contract, 'events')) || []
)
