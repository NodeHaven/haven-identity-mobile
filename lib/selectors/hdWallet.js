// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { createSelector } from 'reselect'
import { get, getIn, count, toJs } from 'mori'

const PATH_VERSION = 7696500
const hdwallet = (state) => state.hdwallet
const identityIndex = (state, index) => index
export const hdAccountIndex = (state, address) => getIn(state.uport, ['byAddress', address, 'hdindex'])
export const hdIdentityIndex = (state, address) => getIn(state.uport, ['byAddress', getIn(state.uport, ['byAddress', address, 'parent']), 'hdindex'])

export const hdRootAddress = createSelector(
  [hdwallet],
  (wallet) => get(wallet, 'root') || undefined
)

export const lastIdentityIndex = createSelector(
  [hdwallet],
  (wallet) => get(wallet, 'identities') ? count(get(wallet, 'identities')) - 1 : undefined
)

export const lastAccountIndex = createSelector(
  [hdwallet, identityIndex],
  (wallet, index) => get(wallet, 'identities') ? getIn(wallet, ['identities', index]) : undefined
)

export const hdPathFor = createSelector(
  [hdIdentityIndex, hdAccountIndex],
  (id, acct) => acct !== null ? `m/${PATH_VERSION}'/${id || 0}'/${acct}'/0'` : undefined
)

export const seedConfirmedSelector = createSelector(
  [hdwallet],
  (wallet) => get(wallet, 'seedConfirmed')
)

export const accountRiskSentSelector = createSelector(
  [hdwallet],
  (wallet) => get(wallet, 'accountRiskSent')
)

export const seedAddresses = createSelector(
  [hdwallet],
  (wallet) => toJs(get(wallet, 'seedAddresses')) || undefined
)
