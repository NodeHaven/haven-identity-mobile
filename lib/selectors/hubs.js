// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { createSelector } from 'reselect'
import { get, peek, toJs, count } from 'mori'
import { currentAddress } from './identities'

export const hub = (state) => state.hub
export const uport = (state) => state.uport

export const hubHead = createSelector(
  [hub],
  (hub) => get(hub, 'head') || undefined
)

export const nextEvent = createSelector(
  [hub],
  (hub) => peek(get(hub, 'events')) || undefined
)

export const hubQueue = createSelector(
  [hub],
  (hub) => toJs(get(hub, 'events')) || []
)

export const hubQueueLength = createSelector(
  [hub],
  (hub) => count(get(hub, 'events')) || 0
)

export const hubCanQueue = createSelector(
  [hub, currentAddress],
  (hub, address) => !get(hub, 'doNotQueue') && address && address !== 'new'
)

export const snapshot = (state) => ({
  uport: toJs(state.uport),
  hdwallet: toJs(state.hdwallet)
})
