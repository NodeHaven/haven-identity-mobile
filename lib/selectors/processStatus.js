// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { createSelector } from 'reselect'
import { get, toJs, keys, map, assoc, sortBy, remove} from 'mori'

// Return a mori object with fuel, nonce, balance, gasPrice and persona
const processStatus = (state) => state.status
const processHistory = (state) => state.history

const sectionName = (state, section) => section

const networking = (state) => state.networking

export const processSection = createSelector(
  [processStatus, sectionName],
  (status, section) => get(status, section)
)

// Return a boolean representing if the client is currently online or not
export const offline = createSelector(
  [networking],
  (net) => net.online === false
)

// Return a boolean representing if there is a process currently running
export const working = createSelector(
  [processSection],
  (stati) => !!get(stati, 'working')
)

// Return a string referring to the current status of the process running
export const statusMessage = createSelector(
  [processSection],
  (stati) => get(stati, 'message')
)

// Return a string detailing the error that occurred during the attempted process
export const errorMessage = createSelector(
  [processSection],
  (stati) => get(stati, 'error')
)

// Has process failed
export const failed = createSelector(
  [errorMessage],
  (error) => !!error
)

// Return a boolean representing if there is a process currently running
export const completed = createSelector(
  [processSection],
  (stati) => !!get(stati, 'completed')
)

export const allMessages = createSelector(
  [processStatus], (stati) => toJs(sortBy(s => get(s, 'section'), map(section => assoc(get(stati, section), 'section', section), keys(stati))))
)

export const messageHistory = createSelector(
  [processHistory],
  (messages) => toJs(messages) || []
)