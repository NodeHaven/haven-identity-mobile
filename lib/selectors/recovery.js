// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { createSelector } from 'reselect'
import { get } from 'mori'
import { networkSettings } from './chains'

const recoveryState = (state) => state.recovery

// Return a string representing the currently used recoveryType
export const recoveryType = createSelector(
  [networkSettings],
  (settings) => settings.recoveryType
)
// Return the current recovery seed of the current identity
export const seed = createSelector(
  [recoveryState],
  (recovery) => get(recovery, 'seed')
)
// Return the seedWords for the current identity
export const seedWords = createSelector(
  [seed],
  (words) => words ? words.split(/ /) : []
)
// Return the current wordNo from the seed string
export const wordNo = createSelector(
  [recoveryState],
  (recovery) => get(recovery, 'wordNo', 0)
)

// Return the current word from the seed string
export const currentWord = createSelector(
  [seedWords, wordNo],
  (words, i) => words.length > 0 ? words[i] : null
)