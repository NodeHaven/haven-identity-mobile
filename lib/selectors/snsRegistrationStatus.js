// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { createSelector } from 'reselect'

const snsState = (state) => state.sns

// Return the deviceToken for the push notification service
// undefined otherwise
export const deviceToken = createSelector(
  [snsState],
  (sns) => sns.deviceToken
)

// Return the amazon resource name endpoint for the push notification service
// undefined otherwise
export const endpointArn = createSelector(
  [snsState],
  (sns) => sns.endpointArn
)

// Returns an object of the skipped push notifications
// undefined otherwise
export const skippedPushNotifications = createSelector(
  [snsState],
  (sns) => sns.skippedPushNotifications
)
