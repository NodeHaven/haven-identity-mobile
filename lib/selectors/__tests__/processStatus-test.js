// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { working, offline, statusMessage, errorMessage, failed, completed, allMessages, messageHistory } from 'uPortMobile/lib/selectors/processStatus'
import { toClj, hashMap, list } from 'mori'

const empty = {
  networking: { online: true },
  status: hashMap()
}

const offlineEmpty = {
  networking: { online: false },
  status: hashMap()
}

const statusWithMessages = toClj({
  unnu: {
    message: 'Woohoo',
    error: 'nooooh'
  },
  sensui: {
    working: true
  },
  nisaba: {
    completed: true
  }
})

const history = list(
  { section: 'unnu',
    message: 'Woohoo'
  },
  { section: 'unnu',
    error: 'nooooh'
  },
  {
    section: 'sensui',
    message: 'funded'
  }
)

const withMessages = {
  networking: { online: true },
  status: statusWithMessages,
  history
}

const offlineWithMessages = {
  networking: { online: false },
  status: statusWithMessages
}

it('returns true if working', () => {
  expect(working(empty, 'unnu')).toBeFalsy()
  expect(working(withMessages, 'unnu')).toBeFalsy()
  expect(working(withMessages, 'sensui')).toBeTruthy()
  expect(working(offlineEmpty, 'sensui')).toBeFalsy()
  expect(working(offlineWithMessages, 'sensui')).toBeTruthy()
})

it('returns true if working', () => {
  expect(working(empty, 'unnu')).toBeFalsy()
  expect(working(withMessages, 'unnu')).toBeFalsy()
  expect(working(withMessages, 'sensui')).toBeTruthy()
})

it('returns status message', () => {
  expect(statusMessage(empty, 'unnu')).toBeNull()
  expect(statusMessage(withMessages, 'unnu')).toEqual('Woohoo')
  expect(statusMessage(withMessages, 'sensui')).toBeNull()
})

it('returns error message', () => {
  expect(errorMessage(empty, 'unnu')).toBeNull()
  expect(errorMessage(withMessages, 'unnu')).toEqual('nooooh')
  expect(errorMessage(withMessages, 'sensui')).toBeNull()
})

it('returns true if offline', () => {
  expect(offline(empty)).toBeFalsy()
  expect(offline(withMessages)).toBeFalsy()
  expect(offline(offlineEmpty)).toBeTruthy()
  expect(offline(offlineWithMessages)).toBeTruthy()
})

it('returns failed', () => {
  expect(failed(empty, 'unnu')).toBeFalsy()
  expect(failed(withMessages, 'unnu')).toBeTruthy()
  expect(failed(withMessages, 'sensui')).toBeFalsy()
})

it('returns completed', () => {
  expect(completed(empty, 'unnu')).toBeFalsy()
  expect(completed(withMessages, 'unnu')).toBeFalsy()
  expect(completed(withMessages, 'sensui')).toBeFalsy()
  expect(completed(withMessages, 'nisaba')).toBeTruthy()
})

it('allMessages()', () => {
  expect(allMessages(empty)).toMatchSnapshot()
  expect(allMessages(withMessages)).toMatchSnapshot()
})

it('messageHistory()', () => {
  expect(messageHistory(empty)).toMatchSnapshot()
  expect(messageHistory(withMessages)).toMatchSnapshot()
})
