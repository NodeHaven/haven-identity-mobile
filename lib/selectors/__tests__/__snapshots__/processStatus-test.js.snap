// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allMessages() 1`] = `Array []`;

exports[`allMessages() 2`] = `
Array [
  Object {
    "completed": true,
    "section": "nisaba",
  },
  Object {
    "section": "sensui",
    "working": true,
  },
  Object {
    "error": "nooooh",
    "message": "Woohoo",
    "section": "unnu",
  },
]
`;

exports[`messageHistory() 1`] = `Array []`;

exports[`messageHistory() 2`] = `
Array [
  Object {
    "message": "Woohoo",
    "section": "unnu",
  },
  Object {
    "error": "nooooh",
    "section": "unnu",
  },
  Object {
    "message": "funded",
    "section": "sensui",
  },
]
`;
